microcosm_postgres/encryption/models.py

```diff
@@ -0,0 +1,104 @@
+"""
+Encryption-related models.
+
+"""
+from typing import Sequence, Tuple
+
+from sqlalchemy import Column, event, LargeBinary, String
+from sqlalchemy.dialects.postgresql import ARRAY
+
+from microcosm_postgres.encryption.encryptor import Encryptor
+
+
+class EncryptableMixin:
+    """
+    A (conditionally) encryptable model.
+
+    Using SQLAlchemy ORM events to intercept instance construction and loading to
+    enforce encryption (if appropriate for the `encryption_context_key`). Should be
+    combined with database constraints to enforce that the instance is *either* encrypted
+    or un-encrypted, but *not* both.
+
+    Must define:
+
+     -  An `encryption_context_key` property (defaults to `self.key`)
+     -  A settable, `plaintext` property (defaults to `self.value`)
+     -  A settable, `ciphertext` property (not defaulted)
+
+    """
+    __encryption_context_key__ = "key"
+    __plaintext__ = "value"
+
+    @property
+    def encryption_context_key(self) -> str:
+        return getattr(self, self.__encryption_context_key__)
+
+    @property
+    def plaintext(self) -> str:
+        return getattr(self, self.__plaintext__)
+
+    @plaintext.setter
+    def plaintext(self, value: str) -> None:
+        return setattr(self, self.__plaintext__, value)
+
+    @property
+    def ciphertext(self) -> Tuple[bytes, Sequence[str]]:
+        raise NotImplementedError("Encryptable must implement `ciphertext` property")
```

Comment:
why not use `abstractproperty`? because the setter would still require indirect abstraction?

Reply:
As far as I know, yes